### PR TITLE
DYN-6794 python script editor convert legacy tabs to spaces

### PR DESCRIPTION
**Pull request by ivaylo-matov**
### Purpose

PR to address https://jira.autodesk.com/browse/DYN-6794 & https://github.com/DynamoDS/Dynamo/issues/15033 .
Added button in the Python script editor to convert tab indentations from legacy Python code to space indentations.
Includes test.

DYN-6794-fix

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB

### Release Notes

Added button in Python to convert tab indentations to space indentations so that users can run legacy code.
Icon for the convert button is a place holder.

PS: Tried a few options where we suppress the conversion to spaces in PythonIndentationStrategy but ultimately they all lead to problems beyond just pasting code with tab indentations. Preserving the current strategy and converting legacy code to the modern convention seem a better solution.

### Reviewers

@dnenov
@reddyashish

### FYIs

@Amoursol

### CODE REVIEW

File: src/Libraries/PythonNodeModels/Properties/Resources.Designer.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace PythonNodeModels.Properties {
     // class via a tool like ResGen or Visual Studio.
     // To add or remove a member, edit your .ResX file then rerun ResGen
     // with the /str option, or rebuild your VS project.
-    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "16.0.0.0")]
+    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "17.0.0.0")]
```
Comment (zeusongit) [inline src/Libraries/PythonNodeModels/Properties/Resources.Designer.cs:22]: Please revert this to 16

Comment (ivaylo-matov) [reply src/Libraries/PythonNodeModels/Properties/Resources.Designer.cs]: done

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs [hunk 2]
```diff
@@ -483,6 +483,17 @@ private void OnMigrationAssistantClicked(object sender, RoutedEventArgs e)
                 Dynamo.Logging.Categories.PythonOperations);
             NodeModel.RequestCodeMigration(e);
         }
+        private void OnConvertTabsToSpacesClicked(object sender, RoutedEventArgs e)
+        {
+            if (NodeModel == null)
+                throw new NullReferenceException(nameof(NodeModel));
+
+            if (editText.Document != null)
```
Comment (QilongTang) [inline src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs:491]: Would you replace the check here to use String.IsNullOrEmpty()?

File: test/Libraries/DynamoPythonTests/PythonEditTests.cs [hunk 3]
```diff
@@ -68,6 +69,22 @@ private void UpdatePythonNodeContent(ModelBase pythonNode, string value)
             ViewModel.ExecuteCommand(command);
         }
 
+        /// <summary>
+        ///     Counts the non-overlapping occurrences of a specified substring within a given string.
+        /// </summary>
+        private int CountSubstrings(string code, string subscting)
```
Comment (QilongTang) [inline test/Libraries/DynamoPythonTests/PythonEditTests.cs:75]: There is not already a string utility function for this behavior?

**Review by QilongTang — COMMENTED** [review]
LGTM with a few comments

**Review by zeusongit — APPROVED** [review]

File: test/Libraries/DynamoPythonTests/PythonEditTests.cs (hunk 3 above)
Comment (QilongTang) [reply test/Libraries/DynamoPythonTests/PythonEditTests.cs:75]: I looked it up, seems not :)

**Comment by QilongTang** [comment]
Waiting for PR checks before merging

**Comment by QilongTang** [comment]
Regressions:
[DynamoCoreWpfTests.PackageManager.PackageManagerUITests.FiltersContextMenuStaysOpen](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/15975/testReport/junit/DynamoCoreWpfTests.PackageManager/PackageManagerUITests/FiltersContextMenuStaysOpen/)
[DynamoCoreWpfTests.PackageManager.PackageManagerViewExtensionTests.PackageManagerViewExtesion_SendsNotificationForPackagesThatTargetDifferentHost_AtExtensionLoad](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/15975/testReport/junit/DynamoCoreWpfTests.PackageManager/PackageManagerViewExtensionTests/PackageManagerViewExtesion_SendsNotificationForPackagesThatTargetDifferentHost_AtExtensionLoad/)
[DynamoCoreWpfTests.PackageManager.PackageManagerViewExtensionTests.PackageManagerViewExtesion_SendsNotificationForPackagesThatTargetDifferentHost_AtLatePackageLoad](https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/15975/testReport/junit/DynamoCoreWpfTests.PackageManager/PackageManagerViewExtensionTests/PackageManagerViewExtesion_SendsNotificationForPackagesThatTargetDifferentHost_AtLatePackageLoad/)

**Review by reddyashish — APPROVED** [review]
Running the job again.

**Comment by reddyashish** [comment]
https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/15991/testReport/ looks good. Only 1 test GraphNodeManagerViewExtensionTests.ContainsEmptyListOrNullTest has failed which was recently changed to fix the flaky behavior.

**Comment by chuongmep** [comment]
Can I know what happen with the prevous script ? So it will automation or show the issue when user keep use old dynamo python script ?

**Comment by QilongTang** [comment]
> Can I know what happen with the prevous script ? So it will automation or show the issue when user keep use old dynamo python script ? I don't want to see the same headache as Ironpython 2.7 converts to Cpython3, causing the sets to break.

Hi @chuongmep My understanding is that there is no automation here, so if you open a legacy script, it should not be affected without user manually clicking on that new convert button.

**Pull request merged**